### ISSUE TITLE
allow multithreading for lsr

### ISSUE THREAD

**Issue by fJancsoSzabo** [outbrain/zookeepercli#34]
`lsr` is **incredibly** slow with large zookeeper instances when running it from the root.  I ran some benchmarks and it seems like the majority of the time is spent waiting due to network latency.

It should be possible to increase the speed of this by allowing multiple goroutines to query zookeeper.  I'm thinking of adding two flags `--parallelize` and `--numgoroutines`.  Would there be any interest?

**Comment by shlomi-noach** [comment]
Yes, makes perfect sense, thank you. To clarify, do you wish to write this or are you asking this to be written?

**Comment by fJancsoSzabo** [comment]
I'd be happy to write this, but I just wanted to make sure it wouldn't be seen as useless and thrown out.

**Comment by shlomi-noach** [comment]
Cool. Please go ahead.

**Comment by fJancsoSzabo** [comment]
Do you know if the zookeeper library allows multiple concurrent requests per connection?

**Comment by shlomi-noach** [comment]
I do not know offhand

**Comment by fJancsoSzabo** [comment]
Closing due to lack of interest.

**Comment by shlomi-noach** [comment]
Kindly reopen. Sorry for the long wait, huge backlog and my daily job cause
delays.

**Comment by fJancsoSzabo** [comment]
I've found race conditions in the implementation and since changed a lot of code.  I'm happy to open a pr from my [my master](https://github.com/fJancsoSzabo/zkcli) to yours with my code changes, it contains what is basically a rewrite of `go/cmd/zookeepercli.go`.  It has 100% test coverage (within the rewritten portion), addresses #27, #38, #41, this ticket and another bug that's unlikely to ever surface I found in `zk.go`.  I'm planning on eventually reaching 100% overall test coverage.

The downside is **many** backwards-compatibility breaking API changes and more dependencies.  It's the sort of thing that would require a major version change, but if it's friendly I'm more than happy to contribute.  Personally I find it way nicer to use?
